File: sport-log-server/src/model/cardio.rs
use std::io::Write;

use chrono::NaiveDateTime;
use diesel::{
    deserialize,
    pg::Pg,
    serialize::{self, Output, WriteTuple},
    sql_types::{Double, Float, Integer},
    types::{FromSql, Record, ToSql},
};
use diesel_derive_enum::DbEnum;
use serde::{Deserialize, Serialize};

use sport_log_server_derive::{Create, Delete, GetAll, GetById, Update};

use crate::{
    model::{MovementId, UserId},
    schema::{cardio_session, route},
};

#[derive(DbEnum, Debug, Serialize, Deserialize)]
pub enum CardioType {
    Training,
    ActiveRecovery,
    Freetime,
}

#[derive(SqlType, Serialize, Deserialize, Debug)]
#[postgres(type_name = "position")]
pub struct Position {
    longitude: f64,
    latitude: f64,
    elevation: f32,
    time: i32,
}

impl ToSql<Position, Pg> for Position {
    fn to_sql<W: Write>(&self, out: &mut Output<W, Pg>) -> serialize::Result {
        WriteTuple::<(Double, Double, Float, Integer)>::write_tuple(
            &(self.longitude, self.latitude, self.elevation, self.time),
            out,
        )
    }
}

impl FromSql<Position, Pg> for Position {
    fn from_sql(bytes: Option<&[u8]>) -> deserialize::Result<Self> {
        let (longitude, latitude, elevation, time) =
            FromSql::<Record<(Double, Double, Float, Integer)>, Pg>::from_sql(bytes)?;
        Ok(Position {
            longitude,
            latitude,
            elevation,
            time,
        })
    }
}

pub type RouteId = i32;

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "route"]
pub struct Route {
    pub id: RouteId,
    pub user_id: UserId,
    pub name: String,
    pub distance: i32,
    pub ascent: Option<i32>,
    pub descent: Option<i32>,
    pub track: Option<Vec<Position>>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "route"]
pub struct NewRoute {
    pub user_id: UserId,
    pub name: String,
    pub distance: i32,
    pub ascent: Option<i32>,
    pub descent: Option<i32>,
    pub track: Option<Vec<Position>>,
}

pub type CardioSessionId = i32;

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "cardio_session"]
pub struct CardioSession {
    pub id: CardioSessionId,
    pub user_id: UserId,
    pub movement_id: MovementId,
    pub cardio_type: CardioType,
    pub datetime: NaiveDateTime,
    pub distance: Option<i32>,
    pub ascent: Option<i32>,
    pub descent: Option<i32>,
    pub time: Option<i32>,
    pub calories: Option<i32>,
    pub track: Option<Vec<Position>>,
    pub avg_cycles: Option<i32>,
    pub cycles: Option<Vec<f32>>,
    pub avg_heart_rate: Option<i32>,
    pub heart_rate: Option<Vec<f32>>,
    pub route_id: Option<RouteId>,
    pub comments: Option<String>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "cardio_session"]
pub struct NewCardioSession {
    pub user_id: UserId,
    pub movement_id: MovementId,
    pub cardio_type: CardioType,
    pub datetime: NaiveDateTime,
    pub distance: Option<i32>,
    pub ascent: Option<i32>,
    pub descent: Option<i32>,
    pub time: Option<i32>,
    pub calories: Option<i32>,
    pub track: Option<Vec<Position>>,
    pub avg_cycles: Option<i32>,
    pub cycles: Option<Vec<f32>>,
    pub avg_heart_rate: Option<i32>,
    pub heart_rate: Option<Vec<f32>>,
    pub route_id: Option<RouteId>,
    pub comments: Option<String>,
}

File: sport-log-client/lib/database/tables/cardio.dart

import 'package:sport_log/database/table.dart';
import 'package:sport_log/models/cardio/all.dart';

class CardioSessionTable extends Table<CardioSession> {
  @override DbSerializer<CardioSession> get serde => DbCardioSessionSerializer();
  @override String get setupSql => '''
create table cardio_session (
    id integer primary key,
    user_id integer not null,
    movement_id integer not null references movement(id) on delete no action,
    cardio_type integer not null check(cardio_type between 0 and 2),
    datetime text not null default (datetime('now')),
    distance integer check (distance > 0),
    ascent integer check (ascent >= 0),
    descent integer check (descent >= 0),
    time integer check (time > 0), -- seconds
    calories integer check (calories >= 0),
    track blob,
    avg_cadence integer check (avg_cadence > 0), 
    cadence blob, -- = secs since start
    avg_heart_rate integer check (avg_heart_rate > 0),
    heart_rate blob, -- = secs since start
    route_id integer references route(id) on delete set null,
    comments text,
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    is_new integer not null check (is_new in (0, 1)),
    unique (user_id, movement_id, datetime, deleted)
);
  ''';
  @override String get tableName => 'cardio_session';
}

class RouteTable extends Table<Route> {
  @override DbSerializer<Route> get serde => DbRouteSerializer();
  @override String get setupSql => '''
create table route (
    id integer primary key,
    user_id integer not null,
    name text not null check (length(name) >= 2),
    distance integer not null check (distance > 0),
    ascent integer check (ascent >= 0),
    descent integer check (descent >= 0),
    track blob not null,
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    is_new integer not null check (is_new in (0, 1)),
    unique (user_id, name, deleted)
);
  ''';
  @override String get tableName => 'route';
}
